oauth2.0

zymap opened the pull request.

zymap committed:
oauth2.0

EronWright commented on pkg/auth/oauth2.go:
```diff
@@ -0,0 +1,178 @@
+		if err != nil {
+			return err
+		}
+		if o.isCommandLineTools {
+			grant, err = o.store.LoadGrant(o.issuer.Audience)
+			if err != nil {
+				return err
+			}
+		} else {
+			grant, err := flow.Authorize()
+			if err != nil {
+				return err
+			}
+			err = o.store.SaveGrant(o.issuer.Audience, *grant)
+			if err != nil {
+				return err
+			}
+		}
```
I think it would be better to not vary the behavior here with `isCommandLineTools`.   The uniform thing to do here is `LoadGrant` with the assumption that the store has already been prepared.

To use pulsarctl as a library, the application could call `NewAuthenticationOauth2(issuer, store)` and should handle filling the store with grants.  

To use pulsarctl as a standalone tool, it should call `NewAuthenticationOAuth2WithParams` based on the config file.  As seen here.
https://github.com/apache/pulsar-client-go/pull/307/files

EronWright commented on pkg/auth/oauth2.go:
```diff
@@ -0,0 +1,178 @@
+	return p, err
+}
+
+func NewAuthenticationOAuth2WithParams(
```
Let's find a way to not have a `commandlineTool` boolean.

EronWright commented on pkg/oauth2/oauth2.go:
```diff
@@ -0,0 +1,36 @@
+	resourceCmd := cmdutils.NewResourceCmd(
+		"oauth2",
+		"Operations about oauth2",
+		"Login as a user or active a serviceaccount for using oauth2 authentication",
```
```suggestion
		"Login as a user or activate a service account using OAuth 2.0 authentication",
```

EronWright commented on pkg/oauth2/login.go:
```diff
@@ -0,0 +1,149 @@
+
+	fmt.Println(grant.Token.AccessToken)
+
+	store, err := store2.MakeKeyringStore()
```
We should parameterize `MakeKeyringStore` to take the keyring name, since it is probably using `snctl` as a const.  here we would want `pulsarctl`.

zymap replied:
This is the implementation of the MakeKeyringStore: https://github.com/zymap/pulsarctl/pull/8/files#diff-f5d260c645f08004f2e0bbd021abb0d3R27. I change the keyring name to the pulsarctl.

zymap committed:
Support oauth2 authentication for pulsarctl (#233)

Support oauth2 client credential authentication flow for pulsarctl. The pulsarctl will get the access token from an Authorization service to access protected pulsar resource.

zymap committed:
Address comments

zymap committed:
Merge remote-tracking branch 'sn/master' into oauth2

zymap committed:
Fix tests
